RepClient/src/app/order/store/order.effects.ts

```diff
@@ -0,0 +1,104 @@
+import { Injectable } from '@angular/core';
+import { Router } from '@angular/router';
+import { Actions, createEffect, ofType } from '@ngrx/effects';
+import { Action, Store } from '@ngrx/store';
+import * as _ from 'lodash';
+import { map, take } from 'rxjs/operators';
+import * as fromApp from '../../reducers/index';
+import { ItemModel } from '../../shared/model/company-items/item.model';
+import { OrderItemModel } from '../../shared/model/order/order-item.model';
+import * as OrderActions from './order.actions';
+import { selectItemsByCompany } from './order.reducer';
+import * as fromOrder from './order.reducer';
+
+@Injectable()
+export class OrderEffects {
+
+  resetOrderState$ = createEffect(() => this.actions$.pipe(
+    ofType(OrderActions.resetOrder),
+    map(action => selectItemsByCompany.release())
+  ), {dispatch: false});
+
+  navigateToConfirmationPage$ = createEffect(() => this.actions$.pipe(
+    ofType(OrderActions.navigateToConfirmOrder),
+    map((action: Action) => {
+      this.router.navigateByUrl('/order');
+    })
+  ), {dispatch: false});
+
+  addItem$ = createEffect(() => this.actions$.pipe(
+    ofType(OrderActions.addItem),
+    map((action: (Action & { item: ItemModel })) => this.addToOrder(action.item))
+  ));
+
+  removeItems$ = createEffect(() => this.actions$.pipe(
+    ofType(OrderActions.removeItem),
+    map((action: Action & { item: ItemModel }) => this.removeFromOrder(action.item))
+  ));
+
+  constructor(
+    private router: Router,
+    private actions$: Actions,
+    private store: Store<fromApp.State>
+  ) {
+  }
+
+  private removeFromOrder(item: ItemModel): Action {
+    let updatedOrder: OrderItemModel[] = [];
+    let currentState: fromOrder.State;
+
+    this.store.select('order').pipe(take(1)).subscribe((state: fromOrder.State) => { // Synchronously get last state
+      currentState = state;
+    });
+
+    updatedOrder = [..._.cloneDeep(currentState.items)]; // Never mutate state
+
+    if (currentState.items) {
+      const lookupItem: OrderItemModel = updatedOrder.find(i => i.id === item.id); // Check if our item is there
+
+      if (lookupItem) {
+        lookupItem.quantity -= 1;
+        if (lookupItem.quantity <= 0) { // if quantity 0 or less
+          updatedOrder.splice(updatedOrder.indexOf(lookupItem), 1); // Remove item
+        }
```

Comment:
What if quantity, was somehow already `<=0`?

Reply:
Then the item would be removed regardless as it fulfils the condition?

Reply:
Nvm yeah, idk what I was thinking